docs/architecture/adr-016-validator-consensus-key-rotation.md

```diff
@@ -0,0 +1,120 @@
+# ADR 016: Validator Consensus Key Rotation
+
+## Changelog
+
+- 2019 Oct 23: Initial draft
+- 2019 Nov 28: Add key rotation fee
+
+## Context
+
+Validator consensus key rotation feature has been discussed and requested for a long time, for the sake of safer validator 
+key management policy (e.g. https://github.com/tendermint/tendermint/issues/1136). So, we suggest one of the simplest form of
+validator consensus key rotation implementation mostly onto Cosmos-SDK.
+
+## Decision
+
+### Pseudo procedure for consensus key rotation
+
+- create new random consensus key.
+- create and broadcast a transaction with a `MsgRotateConsPubKey` that states the new consensus key is now coupled with the validator operator with signature from the validator's operator key.
+- old consensus key becomes unable to participate on consensus immediately after the update of key mapping state on-chain.
+- start validating with new consensus key.
+- validators using HSM and KMS should update the consensus key in HSM to use the new rotated key after the height `h` when `MsgRotateConsPubKey` committed to the blockchain.
+
+
+### Considerations
+
+- consensus key mapping information management strategy
+    - store history of each key mapping changes in the kvstore.
+    - the state machine can search corresponding consensus key paired with given validator operator for any arbitrary height in a recent unbonding period.
+    - the state machine does not need any historical mapping information which is past more than unbonding period.
+- limits
+    - a validator cannot rotate its consensus key more than `MaxConsPubKeyRotations` time for any unbonding period, to prevent spam.
+    - parameters can be decided by governance and stored in genesis file.
+- key rotation fee
+    - a validator should pay `KeyRotationFee` to rotate the consensus key which is calculated as below
+    - `KeyRotationFee` = `InitialKeyRotationFee` * 2^(number of rotations in `ConsPubKeyRotationHistory` in recent unbonding period)
+- slash module
+    - slash module can search corresponding consensus key for any height so that it can decide which consensus key is supposed to be used for given height.
+- abci.ValidatorUpdate
+    - tendermint already has ability to change a consensus key by ABCI communication(`ValidatorUpdate`).
+    - validator consensus key update can be done via creating new + delete old by change the power to zero.
+    - therefore, we expect we even do not need to change tendermint codebase at all to implement this feature.
+- new genesis parameters
+    - `MaxConsPubKeyRotations` : maximum number of rotation can be executed by a validator in recent unbonding period.
+    - `InitialKeyRotationFee` : the initial key rotation fee when no key rotation has happened in recent unbonding period.
+
+
+### Workflow
```

Comment:
I'm confused about which module will this feature be a part of. I assume staking? Can you add that as a comment?

Reply:
I added two parts as below

1) in Considerations
- new genesis parameters _in `staking` module_

2) in Workflow
_- Note : All above features shall be implemented in `staking` module._